docs/source/advanced/js-interop.mdx

```diff
@@ -0,0 +1,234 @@
+
+### Accessors and Polymorphism
+
+Typically `responseBased` codegen would create companion objects with accessors for polymorphic models. For example:
+
+```kotlin
+public interface Animal {
+  public val __typename: String
+
+  public val species: String
+
+  public companion object {
+    public fun Animal.asLion() = this as? Lion
+
+    public fun Animal.asCat() = this as? Cat
+  }
+}
+```
+
+Unfortunately, `@JsExport` does not support companion objects nor extension functions (see [limitations](#limitations)).
+What's more, `@JsExport` have no runtime type information on JS so it's impossible to tell at runtime if a given
+instance is a `Cat` ot a `Lion`. To check this, use `__typename` and `unsafeCastOrCast`:
+
+```kotlin
+when (animal.__typename) {
+  "Lion" -> animal.unsafeCastOrCast<Lion>()
+  "Cat" -> animal.unsafeCastOrCast<Cat>()
+}
+```
+
+`unsafeCastOrCast`:
+
+* Uses a `as?` cast on non-JS targets
+* Uses `unsafeCast()` ([doc](https://kotlinlang.org/docs/js-interop.html#casts)) on JS. This does not type checking at
+  all. If for some reason, your response doesn't have the expected shape your program will fail.
+
+### Limitations
+
+* `JsExport` is an experimental feature in Kotlin and Apollo Kotlin and may change in future versions.
+* `JsExport` only makes sense on response based codegen since it requires the Kotlin models to have the same shape as
+  the JSON.
+* On JS, is is not possible to check if a `@JsExport` instance implements a given class. If you need polymorphism, you
+  must check `__typename` to determine what interface to use.
+* Extension functions on generated code break when you use this technique since we are casting a raw JS object and not
+  actually instantiating a class.
+* `generateAsInternal = true` does not work with `JsExport`, since the compiler ends up giving the internal modifier
+  precedence, and thus mangling the property names.
+* Custom adapters can only be used when their target types are supported by JS (see the full list
+  of [supported types](https://kotlinlang.org/docs/js-to-kotlin-interop.html#kotlin-types-in-javascript)).
+* Enums do not support `@JsExport` and are generated as `String`. The Kotlin enum is still generated so you can
+  use `safeValueOf()` to get a Kotlin enum from a `String`
+
+## `DynamicJsJsonReader`
+
+If you would prefer to use `operationBased` models, and performance is not _as_ critical, you can
+use `DynamicJsJsonReader`. `DynamicJsJsonReader` works with a javascript object that is already parsed on the JS side.
+
+In JS reading response byte by byte from a byte array like okio is doing incurs a lot of overhead because Kotlin
+uses `Long` indices in its Arrays, and Longs do not have a JS implementation. 
+
+By bypassing this reading, `DynamicJsJsonReader` allows faster reading of responses while still keeping the full Kotlin type information. 
+
+In testing we've seen a ~2x performance boost on JS platforms using this parser vs ~10x with the `JsExport` approach.
```

Comment:
Would be cool to publish the benchmark for that if you have them around.

Reply:
I can run some official benchmark.js benchmarks of `DynamicJsJsonReader` vs `BufferedSourceJsonReader` this afternoon so we have some hard numbers.

Getting a benchmarks for the `jsexport` approach is a little harder because you need to account for the `.json()` call which happens asynchronously. Let me see what I can do about that.

Reply:
Nice!
Kotlin/JS is direly lacking real life benchmarks and I feel this is as close as it gets to one. As Kotlin/JS gets more attention, this would be useful data points for future optimizations.

Reply:
I did some benchmarking using a big polymorphic query result from the Github API.

The idea was to compare:

 * `JSON.parse` + `unsafeCast` (`JsExporrt`)
 * `JSON.parse` + `DynamicJsJSONReader`
 * `BufferedSourceJsonReader`

I ran the tests on Kotlin 1.8.21 on Chrome 112 on my Macbook Pro M1.

The results are actually more dramatic than what we had in the doc:

```
parse with js export 40327.72686447989 ops/sec
parse with js export 68 runs
parse with dnymaic reader 9989.38589840788 ops/sec
parse with dnymaic reader 54 runs
parse with buffer reader 394.15146896515483 ops/sec
parse with buffer reader 63 runs
```

So dynamic is ~25x faster than buffered, and `JsExport` is ~4x faster than Dynamic.

* [The query](https://github.com/baconz/apollo-kotlin-js-performance/blob/benchmark/shared/src/commonMain/graphql/SearchQuery.graphql)
* [The setup](https://github.com/baconz/apollo-kotlin-js-performance/blob/benchmark/shared/src/jsMain/kotlin/com/baconz/apolloperformance/GithubSearch.kt)
 * [The benchmarking code](https://github.com/baconz/apollo-kotlin-js-performance/blob/benchmark/shared/src/commonMain/resources/index.html#L12)

Reply:
I can clean up that repo and publish it as a benchmarking tool for this, or we could include a snapshot of those numbers. Thoughts?

Reply:
For the record, I'm trying to get the deploy preview approved so we can see the doc in real conditions. I'm not too familiar with netlify but we'll get there!

Reply:
I just pushed up the benchmark section! LMK what you think.